flask_hal/link.py

```diff
@@ -0,0 +1,248 @@
+#!/usr/bin/env python
+# encoding: utf-8
+
+"""
+flask_hal.link
+==============
+
+Implements the ``HAL`` Link specification.
+"""
+
+# Standard Libs
+import json
+
+
+VALID_LINK_ATTRS = [
+    'name',
+    'title',
+    'type',
+    'deprecation',
+    'profile',
+    'templated',
+    'hreflang'
+]
+
+
+class Links(object):
+    """Build a collection of ``HAL`` link objects.
+
+    Example:
+        >>> from flask_hal.link import Link, Links
+        >>> l = Links(
+        ...     Link('foo', 'http://foo.com'),
```

Comment:
The `Link` class handles the creation of valid `_links` objects whilst `LInks` handles a collection of `LInk` objects. I did it this way so you can only ever create properly formed links to the spec. As a minimum a link needs a `rel` and `href`, but you could easily define all of other attributes here as well:

``` python
foo_link = Link('foo', '/foo', name='Foo', profile='...', deprecation=True)
bar_link = Link('bar', '/bar', title='Bar')

links = Links(foo_link, bar_link).to_json()
```

You could also pragmatically build the links:

``` python
links = Links()  # the root _links
objs = ['foo', 'bar']
for obj in objs:
    links.append(Link(obj, url_for('obj', obj), title=obj)
links.to_dict()
```

This way each of our objects could define a `hal_link` property allowing us to potentially easily build these links with minimal effort:

``` python
class Foo(db.Model):
    name = db.Unicode()

    @property
    def hal_link(self):
        return Link('foo', url_for('foo', self.id), name=self.name)
```

Then `_links` building starts to get really simple:

``` python
links = Links()  # the root _links
objs = Foo.query.all()
for obj in objs:
    links.append(obj.hal_link)
links.to_dict()
```